### ISSUE TITLE
Semantic highlighting marking a class usage as a constant

### ISSUE THREAD

**Issue by dgmora** [Shopify/ruby-lsp#1128]
### Ruby version

3.2.2

### Code snippet

This may not be the most lightweight way to reproduce it, but it's visible.
<details><summary>Dockerfile</summary>

```Dockerfile
# Use Alpine as the base image
FROM alpine:latest

# Set the working directory
WORKDIR /root

# Install necessary packages
RUN apk add --no-cache \
  git \
  curl \
  build-base \
  zlib-dev \
  openssl-dev \
  readline-dev \
  linux-headers \
  libffi-dev \
  cmake \
  automake \
  autoconf \
  libtool \
  pkgconf \
  gettext-tiny-dev \
  ruby \
  ruby-dev \
  bash \
  yaml-dev

# Clone and build Neovim from source
RUN git clone https://github.com/neovim/neovim.git && \
  cd neovim && \
  make CMAKE_BUILD_TYPE=Release && \
  make install

# Clone the LazyVim starter repository
RUN git clone https://github.com/LazyVim/starter ~/.config/nvim

# Copy the content of custom_plugins.lua from an external file
COPY custom_plugins.lua /root/.config/nvim/lua/plugins/custom_plugins.lua

# Create example.rb inline
RUN echo -e "class Foo\n  def self.bar\n    Foo.new # `Foo` should be the same color in both lines\n  end\nend" > ~/example.rb

# Create Gemfile inline
RUN echo "source 'https://rubygems.org'" > ~/Gemfile

# Install bundler and gems specified in the Gemfile
RUN gem install bundler && \
  bundle install

# Start Neovim headlessly and install plugins
RUN nvim -es -u ~/.config/nvim/init.lua -i NONE -c "Lazy" -c "qa"

# Start a shell when the Docker container is run
CMD ["sh"]
```

</details>

```lua
-- store this in the same folder of the Dockerfile as `custom_plugins.lua`
return {
    {
        "williamboman/mason.nvim",
        opts = function(_, opts)
            vim.list_extend(opts.ensure_installed, {
                "ruby-lsp",
            })
        end,
    },
    {
        "neovim/nvim-lspconfig",
        opts = {
            servers = {
                ruby_ls = {},
            },
        },
    },
}
```

Run:
```bash
docker build -t example .
docker run -ti example
# inside
nvim example.rb
```

Wait a bit and you'll see some notifications like `installing ruby_ls`, `ruby_ls installed`, and some seconds later the server will be started automatically. At that point you'll see the color of the second `Foo` change.

### Description

I wanted to replace solargraph with `ruby_ls` with neovim (and LazyVim). When switching I noticed that the highlighting was different. I've noticed that this is most probably because `ruby_ls` semantic highlighting.

I'm using https://github.com/maxmx03/solarized.nvim with `neo` theme, although I can reproduce it with other themes.
After using `:Inspect`, i've noticed the following

<img width="829" alt="Screenshot 2023-10-20 at 20 30 14" src="https://github.com/Shopify/ruby-lsp/assets/3609616/b79efb49-b757-43f6-a236-930295a4bcee">

`Foo` here is highlighted as a a namespace / constant, where on declaration it's as a class:

<img width="821" alt="Screenshot 2023-10-20 at 20 36 11" src="https://github.com/Shopify/ruby-lsp/assets/3609616/d1619d05-207d-4c90-b2c4-7de91a13db70">

### Expected output

I would expect that in both cases `Foo` is the same color (orange, as a `Type`)

**Comment by vinistock** [comment]
Thank you for the report! Yes, we started out by using constant for every reference before we had an index.

If anyone is interest is looking into this, we now need to pass the index to the semantic highlighting request and check the correct type of constant references to produce the correct semantic token. We also need to update the `SEMANTIC_HIGHLIGHTING.md` file to reflect the changes.

The main concern with this is performance, since every constant reference found in a file will have to be resolved - which can prove very expensive.

**Comment by vinistock** [comment]
Sure, I can add more details. From the parser's perspective (and Ruby's to some degree), there's no difference between a reference to a constant, class or module. They are all just constant references. However, it's useful to differentiate them for developers, since the intent is different - even if under the hood they are the same.

To be able to differentiate them, we need to:

- Know all declarations in the project. This allows us to know if the reference was declared as a class, module or constant
- Resolve the reference according to the current name. Consider this file as an example:

```ruby
module Foo
  module Bar
    class Baz; end

    Baz
  end
end
```
Here we have a reference to `Baz` inside `Foo` and `Bar`. Resolving this reference means figuring out exactly to which constant it points to. Is it `Foo::Bar::Baz`, `Foo::Baz` or `Baz`? All of those are possible from Ruby's standpoint.

So what we need to do to fix this is:

- [the semantic highlighting listener](https://github.com/Shopify/ruby-lsp/blob/4e80438ebac71be1a5cbffa487e47eed2a4bc65e/lib/ruby_lsp/executor.rb#L101) has to receive the index as an argument
- In the [semantic highlighting listener](https://github.com/Shopify/ruby-lsp/blob/main/lib/ruby_lsp/requests/semantic_highlighting.rb), we need to add events for classes and module declarations so that we can keep track of the current nesting. Similar to what we do in the indexing visitor, we need [an array](https://github.com/Shopify/ruby-lsp/blob/4e80438ebac71be1a5cbffa487e47eed2a4bc65e/lib/ruby_indexer/lib/ruby_indexer/visitor.rb#L12) to store the current nesting. In the `on_class_node_enter` event, we push the name of the class to the stack. In the `on_class_node_leave` event, we pop from the stack. Same thing for modules
- Finally, we use the index, the current nesting and the name of the constant reference to invoke resolve [like we do in definition](https://github.com/Shopify/ruby-lsp/blob/4e80438ebac71be1a5cbffa487e47eed2a4bc65e/lib/ruby_lsp/requests/definition.rb#L133), which will return all declarations of that constant. The entries will be of a different class depending on what the constant is. If it's a class, it'll be an `Entry::Class`, if it's a module `Entry::Module` and `Entry::Constant` for constants. Based on that information, we should decide the right token to return to the editor using the [specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokenTypes)
    - Module -> namespace
    - Class -> class
    - Constant -> namespace (the spec doesn't have a type for constants, so this is the best we can do)

I do want to emphasize something though: resolving constant references can be expensive and semantic highlighting runs on every single keypress. Even if this is implemented correctly, we may still not be able to merge it if performance is degraded excessively.

**Comment by vinistock** [comment]
Yes, the index has all of that information about declarations, but that doesn't remove the need for resolution. Take this example:

```ruby
class Baz; end

module Foo
  class Baz; end

  module Bar
    class Baz; end

    Baz
  end
end
```
In this `Baz` reference, you need to figure out if it refers to `Foo::Bar::Baz`, `Foo::Baz` or just `Baz`. It's not enough to just look at the nesting directly, you always need to fully resolve the name first.

It gets even more complicated if you have aliases to modules/classes. For example

```ruby
module Foo
  class Bar; end
end

CONST = Foo

module Baz
  CONST::Bar
end
```

In that case, `CONST::Bar` refers to `Foo::Bar`. But it could also refer to `Baz::CONST::Bar` or even `Baz::Foo::Bar`. Part of the resolution is following aliased namespaces.

And that's why it might become too expensive given that semantic highlighting runs on every keypress.

**Comment by mateusdeap** [comment]
@vinistock Interesting. I'm still going to try and see where I might take this but I'm wondering: Do you know how other languages do it? I'm not sure, but is Ruby unique in this respect? Because if there's something similar in another language, maybe we can look at their LS implementation, if they have one. If not... well... if it does prove too complex and expensive, maybe semantic highlighting won't be a thing hahaha At least until a solution to the performance issue is found.

**Comment by vinistock** [comment]
I don't know of another language where classes, modules and constants are ambiguous. Part of why we need to use the index and resolve references is because there's no way to tell them apart just from syntax in Ruby - it's ambiguous.

If you have

```ruby
Foo::Bar
```
With that information alone, it's not possible to tell what `Foo` or `Bar` are. Both can be a class, a module or a constant alias to another class/module.

If there's another language with the same type of ambiguity, then we can definitely check to see what they are doing.

I do believe that the right implementation is resolving the references to figure out the right token. We can always benchmark to compare the new implementation with the old and understand how much of an impact it has.

**Comment by mateusdeap** [comment]
@vinistock Opened a draft PR. As of next week I won't have as much time available, but I intend to continue working on this. However, if anyone wishes to modify my PR, feel free to do so.
